### PR TITLE
Enhance learning rate in optimisation

### PR DESCRIPTION
**Pull request by jzstark**
This PR is a modification based on Royson's code. It enhances the `gcache` of each iteration during optimisation to contain more than one parameter, and thus enables the implementation of Adam optimiser. It also contains unit test for the `update_ch` and `run` functions of Adam, Adagrad, and RMSprop optimiser.

### CODE REVIEW

**Comment by ryanrhymes** [comment]
OK, I see. Can either of you @jzstark @royson give a brief summary about the changes, e.g. is the PR compatible with all the optimisation functions in this module, or does it only apply to the one for DNN optimiser?

**Comment by jzstark** [comment]
The changes I made are: 
- extending the `State.ch` from `t array array` to `t array array array` to accommodate more than one gcache values. Existing optimisers Adagrad and RMSprop, and signatures of `run` function, are subsequently modified.
- implementation of Adam optimiser, mostly on its `update_ch` and `run` functions.
- minor changes of initial values in `Checkpoint` module.

The changes I made are limited to the `Learning_Rate` module and `Checkpoint` module. It is compatible with all three `minimise` functions. Besides the unittest, I test the Adam and Adagrad optimisers on svm regression (`minimise_weight`), the mnist example (`minimise_network`), and the neural style transfer code (`minimise_function`). They all work as expected. 

My understanding of Royson's code is that it requires modifying the methods of adjusting direction based on learning rate, so both `minimise_weight` and `minimise_function` should also be further included. But maybe @royson can provide more details?

**Comment by royson** [comment]
Ahh. Yes. I tried dividing by the gradient at the `Learning_Rate` module to prevent modifications to the `minimise` functions but it didn't work for me. Perhaps it was due to another bug. Regardless, @jzstark is totally right. Sorry for the suboptimal work. Learning in progress. Thank you!

File: src/base/optimise/owl_optimise_generic.ml [hunk 1]
```diff
@@ -65,14 +65,20 @@ module Make
...
       | Const a          -> fun _ _ _ -> F a
       | Decay (a, k)     -> fun i _ _ -> Maths.(F a / (F 1. + F k * (F (float_of_int i))))
       | Exp_decay (a, k) -> fun i _ _ -> Maths.(F a * exp (neg (F k) * (F (float_of_int i))))
-      | RMSprop (a, _)   -> fun _ _ c -> Maths.(F a / sqrt (c + F 1e-32))
+      | RMSprop (a, _)   -> fun _ _ c -> Maths.(F a / sqrt (c.(0) + F 1e-32))
+      | Adam (a, b1, b2) -> fun i g c -> Maths.(F a *
+          (sqrt (F 1. - F b2 ** F (float_of_int i))) /
+          (F 1. - F b1 ** F (float_of_int i)) *
+          c.(0) / (sqrt c.(1) + F 1e-8) /
+          (g + F 1e-32))
```
[the remark this answers is not in the document]
Comment (jzstark) [reply src/base/optimise/owl_optimise_generic.ml]: Yes, the code at line 81.

[the remark this answers is not in the document]
Comment (ryanrhymes) [reply src/base/optimise/owl_optimise_generic.ml]: Cool, thanks. I will try your NST later, is that zoo gist synced with this PR? I mean, does it use ADAM or do I need to tweak it by myself?

[the remark this answers is not in the document]
Comment (jzstark) [reply src/base/optimise/owl_optimise_generic.ml]: Thanks! The gist uses ADAM optimiser.

**Comment by ryanrhymes** [comment]
This PR is solid. Although the extra division on line 81 introduces some overhead, but this is probably the least intrusive way to introduce ADAM into the optimisation engine, so nice job! 💯

**Pull request merged**